src/OmniSharp.Roslyn/ProjectEventForwarder.cs

```diff
@@ -36,46 +38,47 @@ public void Initialize()
 
         private void OnWorkspaceChanged(object source, WorkspaceChangeEventArgs args)
         {
-            SimpleWorkspaceEvent e = null;
+            SimpleWorkspaceEvent workspaceEvent = null;
 
             switch (args.Kind)
             {
                 case WorkspaceChangeKind.ProjectAdded:
-                    e = new SimpleWorkspaceEvent(args.NewSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectAdded);
+                    workspaceEvent = new SimpleWorkspaceEvent(args.NewSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectAdded);
                     break;
                 case WorkspaceChangeKind.ProjectChanged:
                 case WorkspaceChangeKind.ProjectReloaded:
-                    e = new SimpleWorkspaceEvent(args.NewSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectChanged);
+                    workspaceEvent = new SimpleWorkspaceEvent(args.NewSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectChanged);
                     break;
                 case WorkspaceChangeKind.ProjectRemoved:
-                    e = new SimpleWorkspaceEvent(args.OldSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectRemoved);
+                    workspaceEvent = new SimpleWorkspaceEvent(args.OldSolution.GetProject(args.ProjectId).FilePath, EventTypes.ProjectRemoved);
                     break;
+                default:
+                    return;
             }
 
-            if (e != null)
+            var added = false;
+            lock (_eventLocks.GetOrAdd(workspaceEvent, (_) => { added = true; return new object(); }))
             {
-                lock (_lock)
+                // We are already processing a similar event, no need send it again
+                if (added)
                 {
-                    var removed = _queue.Remove(e);
-                    _queue.Add(e);
-                    if (!removed)
+                    object payload = null;
+
+                    try
                     {
-                        Task.Factory.StartNew(async () =>
+                        if (workspaceEvent.EventType != EventTypes.ProjectRemoved)
                         {
-                            await Task.Delay(500);
-
-                            object payload = null;
-                            if (e.EventType != EventTypes.ProjectRemoved)
-                            {
-                                payload = await GetProjectInformationAsync(e.FileName);
-                            }
+                            // Project information should be up-to-date so there's no need to wait.
+                            // TODO: Use AsyncLock (or SemaphoreSlim) instead of lock to allow async/await.
+                            payload = GetProjectInformationAsync(workspaceEvent.FileName)
```

Comment:
maybe update to SemaphoreSlim(1) straight away, the changeset is not that big yet

Reply:
Yes, I was just concerned with the disposal of it. When and who should dispose? If one is waiting on the ”outside”, we can’t dispose it on the ”inside”.

Reply:
If you use a `SemaphoreSlim` instead of an `object` you can just call `semaphore.Wait()` to hold until it releases and then you don't have to remove the event lock, just call `semaphore.Release(1);`

Reply:
Yes, but `SemaphoreSlim` implements `IDisposable`, thus needs to be disposed, orherwise we’ll get a memory leak here. And I’m not sure it’s safe to dispose it after it’s removed from the dictionary, and released, as there might be another thread waiting on the semaphore above.

Reply:
I think I’ll just copy over the `AsyncLock` implementation from Orleans https://github.com/dotnet/orleans/blob/master/src/Orleans.Core/Async/AsyncLock.cs